x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/CreateEnrollmentToken.java

```diff
@@ -0,0 +1,208 @@
+            throw new UserException(ExitCodes.CONFIG,
+                "Unable to create an enrollment token. Elasticsearch node HTTP layer SSL configuration Keystore contain multiple " +
+                    "PrivateKey entries where the associated certificate is a CA certificate");
+        } else {
+            final CheckedSupplier<String, Exception> createApiKeyRequestBodySupplier = () -> {
+                XContentBuilder xContentBuilder = JsonXContent.contentBuilder();
+                xContentBuilder.startObject()
+                    .field("name", "enrollment_token_API_key_" + UUIDs.base64UUID())
+                    .field("expiration", ENROLL_API_KEY_EXPIRATION_SEC)
+                    .startObject("role_descriptors")
+                    .startObject("create_enrollment_token")
+                    .field("cluster", "[" + ClusterPrivilegeResolver.ENROLL_NODE.name() + "]")
+                    .endObject()
+                    .endObject()
+                    .endObject();
+                return Strings.toString(xContentBuilder);
+            };
+
+            final CheckedSupplier<String, Exception> getHttpInfoRequestBodySupplier = () -> {
+                XContentBuilder xContentBuilder = JsonXContent.contentBuilder();
+                xContentBuilder.startObject()
+                    .endObject();
+                return Strings.toString(xContentBuilder);
+            };
+
+            try {
+                final URL url = new URL(client.getDefaultURL());
+                final HttpResponse httpResponseApiKey = client.execute("POST", createAPIKeyURL(url), user, password,
+                    createApiKeyRequestBodySupplier, is -> responseBuilder(is, terminal));
+                int httpCode = httpResponseApiKey.getHttpStatus();
+
+                if (httpCode != HttpURLConnection.HTTP_OK) {
+                    throw new UserException(ExitCodes.CANT_CREATE,
+                        "Unexpected response code [" + httpCode + "] from calling POST " + createAPIKeyURL(url));
+                }
+
+                final String apiKey = Objects.toString(httpResponseApiKey.getResponseBody().get("api_key"), "");
+                if (apiKey == null || apiKey.isEmpty()) {
+                    throw new UserException(ExitCodes.CANT_CREATE,
+                        "Could not create an api key.");
+                }
+
+                final HttpResponse httpResponseHttp = client.execute("GET", getHttpInfoURL(url), user, password,
+                    getHttpInfoRequestBodySupplier, is -> responseBuilder(is, terminal));
+                httpCode = httpResponseHttp.getHttpStatus();
+
+                if (httpCode != HttpURLConnection.HTTP_OK) {
+                    throw new UserException(ExitCodes.CANT_CREATE,
+                        "Unexpected response code [" + httpCode + "] from calling GET " + getHttpInfoURL(url));
+                }
+
+                final String address = getPublishedAddress(httpResponseHttp.getResponseBody());
```

Comment:
I've considered more closely the publish and bind addresses.
I think we should expose the bind ones too, in addition to the published address.

I think we discussed that ES should bind for HTTP on all addresses that it can (`network.host: _global_`). This way we increase the chances that the enroll endpoint can be reached in cases where there are several network interfaces (eg docker).
BUT the publish address is documented as an advanced setting, and is currently single valued. There are old code comments that we should support multiple, but I think there was never the impetus to have more than one, since this value is used for display/information purposes only.

This is how I think it should work (roughly, haven't considered the small details):
* If there is no `network.publish_host` set (the default) the enroll token would show the bind addresses
* if the admin sets a value for the `publish_host`, before it runs the setup script (or the other cmd line tool) then the enroll token shows only the publish address